File: e6/Inheritence.java
import java.util.ArrayList;
import java.util.List;


public class Inheritence {
    public static void main(String[] args) {
        Dog indigo = new Dog("Indigo","labradoodle");
        Cat louis = new Cat("Louis", "tuxedo cat");
        Cat jellybean = new Cat("Jellybean", "tabby cat");
        List<Animal> animals = new ArrayList<>();
        animals.add(indigo);
        animals.add(louis);
        animals.add(jellybean);
        List<Cat> cats = new ArrayList<>();
        cats.add(louis);
        cats.add(jellybean);
        
        animalSounds(animals);
        // animalSounds(cats); does not work because List<Cat> /= List<Animal> 
        animalSounds1(animals);
        animalSounds1(cats);
    }

    public static void animalSounds(List<Animal> animals) {
        for(int i = 0; i < animals.size(); i++) {
            Animal a = animals.get(i);
            System.out.println(a.getName() + ", who is a " + a.getBreed() + ", makes the sound " + a.sound());
        }
    }

    /*
    Not valid syntax in Java
    public static <T> void animalSounds1(List<T extends Animal> animals) {
        for(int i = 0; i < animals.size(); i++) {
            System.out.println(animals.get(i).sound());
        }
    }*/

    public static void animalSounds1(List<? extends Animal> animals) {
        for(int i = 0; i < animals.size(); i++) {
            Animal a = animals.get(i);
            System.out.println(a.getName() + ", who is a " + a.getBreed() + ", makes the sound " + a.sound());
        }
    }
}

File: e6/Inheritence.cpp
#include <iostream>
#include <vector>
#include <memory>
#include <string>
using namespace std; // used to avoid prefixing elements with std::

class Animal {
protected:
    string name;
    string breed;
public:
    Animal(string n, string b) : name(n), breed(b) {}
    virtual ~Animal() {}

    virtual string sound() const {
        return "Some generic animal sound";
    }

    string getName() const { return name; }
    string getBreed() const { return breed; }
};

class Dog : public Animal {
public:
    Dog(string n, string b) : Animal(n, b) {}

    string sound() const override {
        return "Woof";
    }
};

class Cat : public Animal {
public:
    Cat(string n, string b) : Animal(n, b) {}

    string sound() const override {
        return "Meow";
    }
};

// Concept to check if a type is derived from Animal
template <typename T>
concept IsAnimal = std::is_base_of<Animal, T>::value;

// Function with a constraint that only accepts types derived from Animal
template <IsAnimal T>
void animalSounds(const std::vector<T*>& animals) {
    for (const auto& animal : animals) {
        std::cout << animal->sound() << std::endl;
    }
}

int main() {
    Dog indigo("Indigo", "Labradoodle");
    Cat louis("Louis", "Tuxedo Cat");
    Cat jellybean("Jellybean", "Tabby Cat");

    // Use pointers to Animal objects
    vector<Animal*> animals = {&indigo, &louis, &jellybean};
    vector<Cat*> cats = {&louis, &jellybean};

    // Works for all animals
    animalSounds(animals);
    
    // Works only for Cat pointers (also derived from Animal)
    animalSounds(cats);

    return 0;
}
